backend/common/utils/type_conversion_utils.py

```diff
@@ -17,129 +51,132 @@ def get_dtypes_and_schemas_of_dataframe(dataframe: pd.DataFrame):
-        return {"type": "float32"}
-    if dtype_kind == "f" and not can_cast_to_float32(dtype, array_values):
-        return {"type": "float64"}
+    if (
+        not isinstance(array, np.ndarray)
+        and not isinstance(array, pd.Series)
+        and not isinstance(array, pd.Index)
+        and not hasattr(array, "dtype")
+    ):
+        raise TypeError("Unsupported data type.")
+
+    dtype = array.dtype
+
+    res = _get_type_info_from_dtype(dtype)
+    if res is not None:
+        return res
+
+    if dtype.kind == "O":
+        if dtype.name == "category":
+            # Sometimes CategoricalDType can be encoded as int or float without further fuss.
+            # Do not specify the categories in the schema - let the client-side figure it out
+            # on its own.  Utilize Series.to_numpy() to do casting that handles categorical
+            # NA/NaN (missing or undefined) categories.
+            if dtype.categories.dtype.kind in ["f", "i", "u"]:
+                return (
+                    _get_type_info(array.to_numpy())[0],
+                    {"type": "categorical"},
+                )
+            else:
+                return (np.dtype(str), {"type": "categorical", "categories": dtype.categories.to_list()})
+
+        # all other extension types are str-encoded
+        return (np.dtype(str), {"type": "string"})
+
+    if dtype.kind in ["i", "u"] and can_cast_to_int(array, np.int32):
+        return (np.int32, {"type": "int32"})
+
+    if dtype.kind == "f" and can_cast_to_float32(array):
+        return (np.float32, {"type": "float32"})
 
     raise TypeError(f"Annotations of type {dtype} are unsupported.")
 
 
-def can_cast_to_float32(dtype, array_values):
+def can_cast_to_float32(array: Union[np.ndarray, pd.Series]) -> bool:
     """
-    Optimistically returns True signifying that a type downcast to float32 is possible whenever the incoming type is
-    a float.
-
-    We also handle a special case here where the array is a Series object with integer categorical values AND NaNs.
-    Since NaNs are floating points in numpy, we upcast the integer array to float32 and return True.
+    Returns True if the underlying type is float, of any width.
+    Else, returns False
     """
-
-    if dtype.kind == "f":
-        if not np.can_cast(dtype, np.float32):
-            logging.warning(f"Type {dtype.name} will be converted to 32 bit float and may lose precision.")
-
-        return True
-
-    if dtype.kind == "O" and array_values.hasnans:
+    # Accept any float as long as magnitude can still be represented.
+    if array.dtype.kind == "f":
+        if not np.can_cast(array.dtype, np.float32):
+            logging.warning(f"Type {array.dtype.name} will be converted to 32 bit float and may lose precision.")
         return True
 
     return False
 
 
-def can_cast_to_int32(dtype, array_values=None):
+def can_cast_to_int(array: Union[np.ndarray, pd.Series], target: np.dtype) -> bool:
```

Comment:
In every use of this function, target is `int32`. Is there a reason you parametrized it?